### DISCUSSION TITLE
Proposal: OCA Package with Core, Recognized and Community overlays including governance and rules

### DISCUSSION THREAD

**Discussion by carlyh-micb** [the-human-colossus-foundation/oca-spec#70] [General]
This is a proposal for the governance of the OCA standard and will/could include overlay and capture_base syntax. This has been discussed in DSWG which forwards this as a proposal.

# Proposal for OCA specification governance
Overlays Capture Architecture is a standardized language for describing data schemas.

Overlays Capture Architecture is optimized for overlays, where the community can improve the functionalities of a schema to meet their needs while maintaining interoperability.

OCA_bundles MAY be collected together with an OCA package and they can also exist independently.

![OCA package](https://github.com/carlyh-micb/OCA_package/blob/main/package.png)

Figure: example OCA Package structure
 
There are multiple levels of overlays that can be incorporated into OCA syntax, Community, Recognized and Core.

# OCA Core overlays
The overlays are contained in the oca_bundle object. There should be widest support for core overlays, they do not change frequently as they may introduce breaking changes, and any software which claims to support OCA MUST support all core overlays plus capture_base.
Core overlays are governed by HCF, incorporated into OCA Repo implementation and include the OCA File documentation and support.
OCA core overlays are the only overlays that can appear within the oca_bundle (in addition to capture_base). The SAID of the oca_bundle when calculated only includes capture_base and OCA core overlays. Dependencies also only include capture_base and OCA core overlays of respective oca_bundles.
Overlays within the core are no longer individually versioned. They version together with each bundle specification release.

# Extension overlays
Extension overlays include Recognized overlays and Community overlays.

## Recognized Overlays
Recognized overlays are governed by HCF which hosts the official source of overlay syntax and documentation. 

## Community Overlays
Community overlays are governed by their communities and outside of the governance of the oca-specification. However, when present in an oca package community overlays will have specific requirements. Community overlays should follow the OCA Overlay Design and Documentation Standard as best practice.

Note that Community overlays can be a community of one individual which could be implemented in a single project and addressing a single need.

# Promotion of Overlays
## Promotion from Community to Recognized: 
Community overlays are first developed in the community and supported by specific use cases. Then, when popular enough, community overlays may be proposed to the OCA standard as Recognized Overlays. Recognized overlays are accepted from the community when they follow the OCA Overlay Design and Documentation Standard.

Community overlays are proposed to the DSWG for a change in status. The DSWG seeks input from the Technology Council on the suitability of the overlay. Approval to Recognized overlay is the resonsibility of the DSWG. The DSWG is responsible for the timely addition of the new Recognized overlay specification to the official standard with input from the Technology Council. 

## Promotion from Recognized to Core: 
HCF maintains the official tool repository for working with OCA. They also maintain the official implementation of working with OCA Core (OCA Repo). To become a Core overlay, the Recognized overlay must be supported by the official HCF OCA Repo and follow the OCA Overlay Documentation Standard.

Community overlays are proposed to the DSWG for a change in status. The DSWG together with the Technology Council approves the acceptance of the overlay as a Core overlay. The DSWG is responsible for the timely addition of the new Core overlay specification to the official standard with required approval from the Technology Council. 

# OCA Standard Documentation
The HCF MUST maintain the official standard specification for oca_bundle, oca_package and OCA Overlay Design and Documentation Standard. OCA Core overlays and Recognized overlays MUST follow the OCA Overlay Design and Documentation Standard. Community overlays MAY follow the OCA Overlay Design and Documentation Standard.

# OCA Capture Base 
TODO Description of syntax of capture base.

# OCA Core Overlays
TODO Description of syntax of Core overlays following the OCA Overlay Design and Documentation Standard.

# OCA Recognized Overlays
TODO Description of syntax of Recognized overlays following OCA Overlay Design and Documentation Standard.

Note: SAID calculations will reference official version 0.9 release of ACDC, but also include a non-normative and helpful description of how SAIDs are calculated as well as reference libraries that will calculate SAIDs.

# OCA Overlay Design and Documentation Standard
The documentation standard outlines the different sections of documentation for each overlay. Core and Required overlays MUST follow this documentation standard, Community overlays SHOULD use this documentation standard. Test cases for each overlay MAY be documented in a separate section collecting all test cases together. OCA File Syntax MAY be documented in a separate section collecting all OCA File syntax together.

1.	Description (MUST: Core and Recognized):
 - Summarize the functionality of the overlay, what needs the overlay is addressing and descriptions of components of the overlay. Any references to community standards belong in this section.
3.	OCA File description (MUST: Core, MAY: Recognized): 
 - If there is an implementation of OCA File syntax used to generate the overlay include a description of the syntax here.
 - Add meta-grammar rules about OCA File here (if OCA File is present MUST: Core and Recognized).
4.	Example (MUST: Core and Recognized): 
 - Provide at least one example of the overlay in fully canonicalized, serialized (in JSON) format including a correctly calculated SAID value. 
 - The capture_base SAID does not need to reference a specific capture_base but MUST be well formed. 
 - The capture_base example SAID is used in the calculations of the overlay SAID. 
 - The example MAY be formatted with line breaks and indentations for readability.
 - Core overlays MUST use Type= “core/overlay/name/” where name is the name of the overlay.
 - Recognized overlays MUST use Type= “recognized/overlay/name/vX.X where name is the name of the overlay and versioning MUST follow semantic versioning recommendations.
 - Community overlays MUST use Type= “community/overlay/name/vX.X” where name is the name of the overlay and versioning MUST follow Semantic Versioning recommendations when part of the OCA Package.
5.	OCA File source example (MUST: Core, MAY: Recognized): 
 - For at least one overlay example given provide the complete OCA File syntax.
6.	Canonicalization rules (MUST: Core and Recognized): (TODO: adapt this content)
 - Canonicalization rules MUST be documented for all objects.
 - All sections of the overlay must be explicitly specified if they are insertion ordered or lexigraphical ordered. Lexicographical sorting is based on the character Unicode values in ascending order, starting with the first character and proceeding through the string.
 - Overlays MUST have at the beginning: SAID (d), type, optional language, then capture base. This order is insertion ordered. `"d"="", "type"="", "language"="" (optional), "capture_base"=""`
 - If multiple languages are present within an overlay object, each language overlay MUST be ordered lexigraphically.
 - After the  properties that would apply to the entire overlay (or data). This order MUST be lexigraphical ordered.
 - If an attributes object is present it MUST follow lexigraphical ordering.
7.	Rules summary (MUST: Core and Recognized): 
 - Summarize all the requirements (MUST and MAY) for the overlay.
8.	Test case: (MUST: Core and Recognized): 
 - Provide at least one fully worked example that (MUST: Core, MAY (MUST?): Recognized) include OCA File. 
 - The worked example MUST be a fully canonicalized, JSON serialized schema bundle with SAIDs calculated. 
 - The fully worked example MUST include a capture_base and any other overlays that the documented overlay depends on. 
 - The example MUST exclude any transformations made for readability that would interfere with the reproducible calculation of the SAID.

# OCA Bundle 
TODO Description of syntax of OCA Bundle

# OCA Package
TODO Description of syntax of OCA Package

**Comment by mitfik** [comment]
Thanks @carlyh-micb for the input. 

I see here few items which are mixed and could be difficult to discuss them at once:

1) Governance of spec and development - this is actually in [place](https://github.com/the-human-colossus-foundation/oca-spec?tab=readme-ov-file#how-to-contribute), keep in mind that we are working with [dominant design](https://en.wikipedia.org/wiki/Dominant_design#:~:text=A%20dominant%20design%20is%20the,to%20command%20significant%20market%20following.) in mind, we build, test and then spec comes out of it not other way around.
2) [Community overlays](https://github.com/the-human-colossus-foundation/oca-spec/discussions/68) - this is exactly why ORFg was created and we are working there on that topic to find out what would be best approach, please follow the #68 discussion to address it there
3) Structure of the new design - this suppose to be address as well within ORFg where solving point 2 would give a clue how to shape new structure to make it maintainable and scalable solution.

For clarity I would suggest to address those problems one by one instead of mixing them.

Since point 1 is already addressed, point 2 and 3 is discussed in other places would close this discussion as duplicate so we can focus on specific points in corresponding place. Feel free to reopen this discussion if you feel is still needed.

**Comment by carlyh-micb** [comment]
I have reopened the discussion because I would like some clarification.

So to clarify - the only official thing that matters first in OCA is the code. After it has been coded, then it is documented. The OCA-spec documentation comes out of what has been coded.

After an RFC has been proposed, it can only move forward into the specification if code to support it has been contributed to the HCF repository? (which one?). Where is the code demonstrated? Do people need to fork an official code repository of HCF and make their own demonstration? After the demonstration is working, then code can be proposed as PR? Who officially ADOPTS the the PR and merges it into the codebase? Is there a version cycle of the release of the code? The person who officially ADOPTS the code is the person(s) who governs the entire specification? Is there any clarity on who that is, how they can make decisions on behalf of the community?

RFCs seem to be stuck at the level of PROPOSED. No one else except Robert and Michal (HCF) have been successful at contributing code and therefore moving forward the specification.

I want to keep this discussion open - this proposal above was discussed at the DSWG and we are seeking clarity and a way to move forward.


For reference:

RFC lifecycle
PROPOSED

To propose an RFC, use these instructions to raise a PR against the repo. Proposed RFCs are considered a "work in progress", even after they are merged. In other words, they haven't been endorsed by the community yet, but they seem like reasonable ideas worth exploring.

DEMONSTRATED

Demonstrated RFCs have one or more implementations available, listed in the "Implementations" section of the RFC document. As with the PROPOSED status, demonstrated RFCs haven't been endorsed by the community, but the ideas put forth have been more thoroughly explored through the implementation(s). The demonstrated status is an optional step in the lifecycle.

ACCEPTED

To get an RFC accepted, build consensus for your RFC on chat and in community meetings. An accepted RFC is incubating on a standards track; the community has decided to polish it and is exploring or pursuing implementation.

ADOPTED
Adopted is acknowledgment of a de facto standard and would be incorporated into the OCA specification. After that process the RFCs would be archived.

**Comment by mitfik** [comment]
> After an RFC has been proposed, it can only move forward into the specification if code to support it has been contributed to the HCF repository? (which one?). 

It can be any repository and any code base, does not have to be HCF reference implementation. The purpose of that is just to demonstrated that propose idea works and make sense, so other can start adopting it. If you like whoever is interested in developing upon given RFC he needs to provide reference implementation so others can use it and adopt.

If there is consensus within the community (after demonstrating it that it works) community can refine the proposal and make sure that it fit into overall ecosystem. Then PR can be created to suggest changes to the spec - can be done by anyone but approval is done via Tc (Technology Council) right now since Tc is responsible for maintenance and fit of the spec into whole DDE ecosystem.

**Reply by carlyh-micb** [reply]
So, the community can develop their overlays and use them within their own test objects of an OCA schema (they need to test it so it needs to go somewhere right)?
After anyone in the community develops their specific implementation and tests it however they want, the OCA Community can evaluate what a community member has created and if it is something that they value and the OCA community may ultimately chose to recognize it as having value?
Only after the community recognizes the value of the overlay, then they can rework the code put in a PR and request that this recognized overlay be moved over to the core OCA specification? And the core specification of OCA is only decided by the TC.

Do I have this right?

**Comment by mitfik** [comment]
> The person who officially ADOPTS the code is the person(s) who governs the entire specification? Is there any clarity on who that is, how they can make decisions on behalf of the community?

We use dominant design approach means that as soon as there is enough adoption we take it from there. Maybe not the most transparent way to do stuff but enforcing people doing stuff because we wrote about it is not good either. 

Generally Tc is responsible for keeping the spec align through the reference implementation means the spec needs to make sense in the context of DDE. Due to that Tc would enforce some implementation choices. E.g. SAID over multihash. Does not mean that you cannot bring changes to those implementation aspect but you need to bring solid argumentation which would fit into DDE scope to get that approval.

**Comment by mitfik** [comment]
> So, the community can develop their overlays and use them within their own test objects of an OCA schema (they need to test it so it needs to go somewhere right)?

Currently that is the case, this is why we want to come up with the process to unify those efforts to have `community overlays` and clear process how they can be created. In the first design we didn't thought about that.


> After anyone in the community develops their specific implementation and tests it however they want, the OCA Community can evaluate what a community member has created and if it is something that they value and the OCA community may ultimately chose to recognize it as having value?

Correct, that is ADOPTION phase. It does not mean that it needs to be done by one individual or someone who created the RFC. It can be anyone who likes that idea and want to experiment with it.

> Only after the community recognizes the value of the overlay, then they can rework the code put in a PR and request that this recognized overlay be moved over to the core OCA specification? And the core specification of OCA is only decided by the TC.

Yes, Tc govern the spec with the input of the community. Bigger push from the community more likely it would be accepted. The reason why Tc is because Tc is tasked by HCF to maint and develop components of DDE in coherent state. E.g. OCA is just one of the component but it needs to fit the piece in authentication layer. Someone who would work on semantic part only does not necessary needs to understand how the authentication part would work so we are not expecting to develop semantic piece which is align with whole DDE but this is where Tc kicks in and making sure that everything fits together.

**Reply by carlyh-micb** [reply]
Ok, so to confirm - OCA exists as a dependency of the DDE (dynamic data economy). As OCA is an important component, any changes to OCA must serve its utility in the DDE. Other people can use OCA and not other parts of the DDE, but any changes to the core specification of OCA will be done to ultimately serve DDE and that is under the direction of the TC.

**Reply by mitfik** [reply]
OCA is core component of the DDE to address Decentralized Semantic. HCF mission is to develop components of DDE which OCA is one of them. Any changes proposed to OCA which would harm DDE won't be accepted. Anyone can use any of the components of DDE separately or in any combination according to the EUPL license. Yes currently OCA spec is under jurisdiction of Technology Council.

**Comment by pknowl** [comment]
@mitfik I believe the current OCA spec should fall under the jurisdiction of the Advisory Council, as it aligns more with data management than with technical architecture. In my view, the Technology Council should focus on the deeper technical components of the architecture (e.g., OCA Bin, OCAFile, OCA Repo, etc.), where the core technology resides.

Regarding OCA JSON files and overlay types, these are primarily driven by data management needs and are no longer highly technical in nature. As the inventor of OCA, I feel strongly that the tech side should remain separate from use case discussions centered around data management and the proposed overlay types that emerge from them.

**Reply by blelump** [reply]
Can we try to keep this space free from Foundation-related discussions?

**Comment by blelump** [comment]
When we considered the early concepts of how to structure the Bundle, we considered various types of "bags" to distinguish the overlays based on multiple characteristics, i.e., origin, purpose of use, etc. However, this presumes an implicit governance pre-baked into the Bundle and that distinct jurisdictions must be physically separated, i.e., via distinguishing core vs. community. This is what #74 discusses under the `provenance` section. Specifically, it's the role of different layers to solve this problem of who we trust in the use case. Core vs. community is precisely about trust that inherently starts with governance. `core` is merely what a given use case considers as `core` and the same with community overlays. Therefore, there's no need for such distinctions, and Bundle is just its capture base and overlays. The use case governance decides essential aspects to address business requirements later mapped onto technology, i.e., by employing overlay A, B, and C, but not X. It doesn't matter what the origin of A, B, and C (whether it's core or extension) because the use case must ensure proper support for all of them.

Assume A and B are from the current core and C is the community overlay. Technically, handling the business process within the use case may employ HCF libraries to handle A and B. However, what #74 discusses under the `provenance` section does not require or impose such an approach. It says that overlays are equal in this sense, and the governance decides what to use, i.e., via an organization's reputation. A piece of code `oca_lib.load(oca_bundle)` is executable with any Bundle, enabling further consumption of the Bundle attributes and their characteristics (delivered through overlays).

The Bundle with no distinction between core vs. community (or anything else) not only must not change its mechanical structure because it fits the current Bundle structure, but it also relies much more heavily upon distributed governance and the autonomy to decide what to use.
